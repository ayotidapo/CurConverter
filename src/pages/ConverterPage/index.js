import { useState, useEffect } from 'react';

import { useSelector, useDispatch } from 'react-redux';
import { convertCurrency } from 'redux/actions/converterAction';
import Loader from 'components/Loader';
import styled from 'styled-components';
import { formatAmount } from 'helpers/functions';
import Card from 'components/Card';
import Input from 'components/Input';
import { useFormik } from 'formik';
import { Link } from 'react-router-dom';
import currencies from 'helpers/currencies';

const PageWrapper = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  padding: 20px;
  align-items: center;

  input:nth-last-of-type(1) {
    padding-left: 60px !important;
  }

  section {
    font-size: 1.4rem;
    width: 90%;
    border: 1px solid #fafafa;
    box-shadow: 0px 40px 30px rgba(0, 0, 0, 0.02);
  }
  ${Card} {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(320px, 1fr));
    grid-gap: 10px 20px;
    margin-bottom: 20px;
  }
  .link {
    margin-top: 40px;
    text-decoration: none;
    color: var(--pri-color);
    font-size: 1rem;
  }
  .result-div {
    margin-left: 20px;
    padding: 10px;
    width: 300px;
  }
  .equals {
    margin-top: 10px;
    color: #000;
  }
`;

const ConverterPage = () => {
  const [inputValue, setInputValue] = useState(' ');
  const [prevData, setPrevData] = useState({});
  const [quoteValue, setQuoteValue] = useState({});
  const { converting } = useSelector((state) => state.converterState);

  const { prevQuote, prevBase, prevAmount, prevResult } = prevData;

  const dispatch = useDispatch();

  const formik = useFormik({
    initialValues: {
      amount: '',
      base: '',
      quote: '',
    },
  });

  const { values, handleChange, handleBlur } = formik;
  const { amount, base, quote } = values;

  const debAmount = amount;

  useEffect(() => {
    if (!localStorage.lConvert) return;
    const lsData = JSON.parse(localStorage.lConvert);
    setPrevData(lsData);
    const formated = debAmount || formatAmount(lsData?.prevAmount);
    setInputValue(formated);
  }, []);

  useEffect(() => {
    const Amt = debAmount || prevAmount;
    const bs = base || prevBase;
    const qt = quote || prevQuote;

    if (Amt && bs && qt) makeConversion(bs, qt);
  }, [debAmount, base, quote]);

  const makeConversion = async (bas, quot) => {
    const { success, data } = await convertCurrency(bas, quot)(dispatch);
    if (!success) return;

    const rate = data?.result;
    const theQuot = quote || prevQuote;
    const quoteAmount = rate[theQuot];
    const theAmt = debAmount || prevAmount;
    const result = quoteAmount * parseFloat(theAmt);
    setQuoteValue({ result, cur: quote });

    const lsData = {
      prevQuote: theQuot,
      prevBase: base || prevBase,
      prevAmount: theAmt,
      prevResult: result,
    };
    localStorage.lConvert = JSON.stringify(lsData);
  };

  return (
    <PageWrapper>
      <section>
        <Card>
          <Input
            id="amount"
            name="amount"
            label="Enter Amount"
            type="text"
            value={inputValue}
            onChange={(e) => {
              handleChange(e);
              const val = e.target.value;
              const newVal = val.replace(/,/g, '');
              setInputValue(newVal);
            }}
            onBlur={(e) => {
              handleBlur(e);
              const val = e.target.value || prevAmount;
              const formatted = formatAmount(val);
              setInputValue(formatted);
            }}
            curSign={base || prevBase}
            testId="amount-div"
          />

          <Input
            data={currencies}
            defaultValue={base || prevBase}
            id="base"
            name="base"
            label="Change From"
            onChange={handleChange}
            type="select"
            testId="base-select"
          />
          <Input
            data={currencies}
            defaultValue={quote || prevQuote}
            id="quote"
            name="quote"
            label="to"
            onChange={handleChange}
            type="select"
            testId="quote-select"
          />
        </Card>

        {converting && <Loader data-testid="loader-div" />}
        {(quoteValue?.result || prevQuote) && !converting && (
          <div className="result-div" data-testid="result-div">
            <span className="pri-col">{base || prevBase}</span>{' '}
            {formatAmount(debAmount || prevAmount)}{' '}
            <small>is equivalent to</small>
            <h1 className="equals">
              <span className="pri-col">
                {quoteValue?.cur || prevQuote} &nbsp;
              </span>
              {formatAmount(quoteValue?.result || prevResult)}
            </h1>
          </div>
        )}
      </section>

      {(base || prevBase) && (
        <Link
          to={{
            pathname: '/current-rates',
            state: {
              base: base || prevBase,
            },
          }}
          className="link"
        >
          View All Exchange Rates
        </Link>
      )}
    </PageWrapper>
  );
};

export default ConverterPage;
