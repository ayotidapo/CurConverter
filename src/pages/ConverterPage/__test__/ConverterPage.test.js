/* eslint-disable no-undef */
/* eslint-disable import/prefer-default-export */
import React from 'react';
import axiosMock from 'axios';
import { renderWithRedux } from 'testHelper';
import { fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import ConverterPage from '../index';

describe('how Converter Page component works', () => {
  afterEach(cleanup);
  beforeEach(() => {
    axiosMock.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          base: 'USD',
          result: {
            NGN: 414,
          },
          updated: '2021-10-24 02:22:15',
          ms: 3,
        },
      }),
    );
  });
  it('renders the component without crash and takes snapshot', async () => {
    const { asFragment } = renderWithRedux(<ConverterPage />);

    expect(asFragment).toMatchSnapshot();
    expect(axiosMock.get).toHaveBeenCalledTimes(0);
  });

  it('renders the component', async () => {
    const { getByTestId, getByRole, store } = renderWithRedux(
      <ConverterPage />,
    );

    let { converterState } = store.getState();

    const amountInput = getByRole('textbox', { name: 'Enter Amount' });
    const baseSelectInput = getByTestId('base-select');
    const quoteSelectInput = getByTestId('quote-select');

    fireEvent.change(amountInput, { target: { value: '500' } });
    fireEvent.change(baseSelectInput, { target: { value: 'USD' } });
    fireEvent.change(quoteSelectInput, { target: { value: 'NGN' } });

    converterState = await store.getState().converterState;
    console.log(converterState);
    expect(converterState.converting).toBe(true);

    expect(axiosMock.get).toHaveBeenCalledWith(
      `/fetch-one?from=USD&to=NGN&api_key=${process.env.REACT_APP_CUR_CONV_KEY}`,
    );
    await waitFor(() => {
      expect(getByText('is equivalent to')).toBeInTheDocument();
    });
    await store.getState().auth;
    expect(converterState.converting).toBe(false);
  });
});
