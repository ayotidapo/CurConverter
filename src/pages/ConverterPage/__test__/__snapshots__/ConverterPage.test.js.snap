// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`how Converter Page component works renders the component without crash and takes snapshot 1`] = `[Function]`;
